pkg/profile/artifact.go

```diff
@@ -140,52 +133,63 @@ func (p *Profile) getResourceIfExists(ctx context.Context, res client.Object) (b
 // MakeArtifacts creates and returns a slice of runtime.Object values, which if
 // applied to a cluster would deploy the profile as a HelmRelease.
 func (p *Profile) MakeArtifacts() ([]runtime.Object, error) {
-	objs := []runtime.Object{}
-	gitRes, helmResources, kustomizeResources, err := p.makeArtifacts()
+	objs, err := p.makeArtifacts()
 	if err != nil {
 		return nil, err
 	}
-
-	objs = append(objs, gitRes)
-	for _, helmRes := range helmResources {
-		objs = append(objs, helmRes)
-	}
-	for _, kustomizeRes := range kustomizeResources {
-		objs = append(objs, kustomizeRes)
-	}
 	return objs, nil
 }
 
-func (p *Profile) makeArtifacts() (*sourcev1.GitRepository, []*helmv2.HelmRelease, []*kustomizev1.Kustomization, error) {
-	var helmResources []*helmv2.HelmRelease
-	var kustomizeResources []*kustomizev1.Kustomization
+func (p *Profile) makeArtifacts() ([]runtime.Object, error) {
+	objs := make([]runtime.Object, 0)
+	var gitRes *sourcev1.GitRepository
 
 	for _, artifact := range p.definition.Spec.Artifacts {
 		switch artifact.Kind {
-		case "HelmChart":
+		case profilesv1.HelmChartKind:
 			helmRes, err := p.makeHelmRelease(artifact)
 			if err != nil {
-				return nil, nil, nil, fmt.Errorf("failed to create HelmRelease resource: %w", err)
-
+				return nil, fmt.Errorf("failed to create HelmRelease resource: %w", err)
 			}
-			helmResources = append(helmResources, helmRes)
-		case "Kustomize":
+			objs = append(objs, helmRes)
+			if artifact.Path != "" && gitRes == nil {
+				// this resource is added at the end because it's generated once.
+				gitRes, err = p.makeGitRepository()
+				if err != nil {
+					return nil, fmt.Errorf("failed to create GitRepository resource: %w", err)
+				}
+			} else if artifact.Chart != nil {
+				helmRep, err := p.makeHelmRepository(artifact.Chart.HelmURL, artifact.Chart.HelmChart)
+				if err != nil {
+					return nil, fmt.Errorf("failed to create HelmRepository resource: %w", err)
+				}
+				objs = append(objs, helmRep)
+			}
```

Comment:
shall we add some tests + validation checks for ensuring `chart` and `path` are not both set?

Reply:
Definitely. I'm actually writing that just now. :) And increasing coverage as well.

Reply:
It'd be nice to see this done via a `Validate()` method on the ProfileDefinition?

Something like...

```
import "knative.dev/pkg/apis"

(p ProfileDefinition) Validate() *apis.Fielderror {
...
}

It doesn't _have_ to be the knative apis.FieldError...but it does have nice functionality for indicating which fields within the resource are invalid.

https://pkg.go.dev/knative.dev/pkg/apis#FieldError

Reply:
I like how we are thinking similarly. :) The only difference is that I actually put the Validate on the Artifact. Are you okay with that @bigkevmcd?

Reply:
Makes sense for it to be at the level of the thing being validated :-)

Reply:
For comparison, the output of using the standard errors for this:

```go
func (in Artifact) Validate() error {
    if in.Chart != nil && in.Path != "" {
      return apis.ErrMultipleOneOf("chart", "path")
    }
    return nil
 }
```

Would get you

`validation failed on artifact helmChartArtifactName1: expected exactly one, got both: chart, path"`
`validation failed on artifact helmChartArtifactName1: chart and path cannot both be defined"`

Reply:
Added error. But I'm wrapping it out, because I like the context of the chart name, if that is okay.